### ISSUE TITLE
Add support for salting blocking

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#527]
See [this discussion](https://github.com/moj-analytical-services/splink/discussions/254)

In a nutshell, we want the settings object to have an optional key of `salt_partitions`, defaulting to None (no salting).

If set, we want the `blocking_rules_to_generate_predictions`, (and `blocking_rule_for_training` in the `EMTrainingSession`) to be transformed as follows.

First, we need a new columns in `__splink__df_concat_with_tf` called (e.g.) `__splink_salt`, which is populated with a uniformly distributed integer between `[1, salt_partitions`.

Supposing 
```
blocking_rules_to_generate_predictions = [
"l.col_a = r.col_a", 
"l.col_b = r.col_b and l.col_c and r.col_c"
]
```

then we want this to be translated into:

```
blocking_rules_to_generate_predictions = [
"l.col_a = r.col_a and l.__splink_salt = 1",
"l.col_a = r.col_a and l.__splink_salt = 2",
...
"l.col_a = r.col_a and l.__splink_salt = salt_partitions", 

"l.col_b = r.col_b and l.col_c and r.col_c and l.__splink_salt = 1"
"l.col_b = r.col_b and l.col_c and r.col_c and l.__splink_salt = 2"
...
"l.col_b = r.col_b and l.col_c and r.col_c and l.__splink_salt = salt_partitions"
]
```

Then everything else can proceed as normal.

Tricky parts:

1. Generating the splink salt.  Syntax likely to be different between backends.  Probably need a similar approach to how we take a random sample e.g. [here](https://github.com/moj-analytical-services/splink/blob/f15cf1af63ede67d6a0b193d2384eb397ca7e72c/splink/spark/spark_linker.py#L191)
2. Need to check this doesn't interact in a weird way with other parts of Splink e.g. the real time linking functions 

Why do we need this?

**Comment by RobinL** [comment]
## Why might we want this?

Some of the most expensive computation in Splink are:

1. Blocking data, because it creates a large number of comparisons
2. Computing the value of the comparison vectors, because some of the case when statements are expensive


We want to evenly distribute these computations across the cluster i.e. each machine does a similar number of computations


A fundamental problem in with data linking is that Spark is not very good at distributing blocking rules, because they are often skewed.

This is the ‘john smith’ problem - that 50k comparisons may be generated within the cartesian product of ‘john smiths’, even though only 10 comparisons are generated within Robin Linacre
So blocking serves to severely accentuate skew (because original skew becomes n^2 skew) 

Spark hasn’t traditionally been very good at dealing with skew, but with adaptive query execution (ADE) it’s getting better.
The problem is that ADE is designed to work with ‘normal’ skew (n), not the type of n^2 skew that comes out of blocking.
So turning on ADE might give a small speed boost but it’s not a magic bullet 

So we have to manually intervene, and tell Splink to re-distrbute the workload between machines at the earliest opportunity (i.e. redistribute the 50k john smiths between machines)
In practice, there are two places we can do it:
1 After blocking, but before computing the comparison vectors
2 After computing the comparison vectors
The problem with (1) is you potentially have to checkpoint (persist to disk) twice, which is expensive
The problem with (2) is that if you don’t redistribute (repartition) after blocking, then all the john smith comparison vectors end up being computed on a single machine (edited) 


There’s isn’t an obvious best solution, I’m currently experimenting with (1) and (2) to see what gives the fastest result in practice


Note this is particularly an issue for EM iterations, because in EM we want to computer the comparison vectors once, and iteratively re-use them for EM



It may be that we want to build salting into Splink, to try and avoid skew in the first place, which would mean we wouldn’t need to worry so much about (1) skew after blocking and could focus on (2)

**Comment by Th368MoJ** [comment]
On this discussion, can we not improve the load balancing further by evaluating how many comparisons will be created by our various blocking rules and then differing our salting allocations accordingly?

So, if we had say:
```
blocking_rules_to_generate_predictions = [
"l.col_a = r.col_a", 
"l.col_b = r.col_b and l.col_c and r.col_c"
]
```

and we knew that `"l.col_a = r.col_a"` created 50,000 comparisons, whereas `"l.col_b = r.col_b and l.col_c and r.col_c"` only created 10,000, then we could distribute our salt partitions like so, to balance the load:
```
"l.col_a = r.col_a and __splink_salt = 1",
...
"l.col_a = r.col_a and __splink_salt = 10",
"l.col_b = r.col_b and l.col_c and r.col_c and __splink_salt = 1",
"l.col_b = r.col_b and l.col_c and r.col_c and __splink_salt = 2"
```

**Comment by ericmanning** [comment]
> On this discussion, can we not improve the load balancing further by evaluating how many comparisons will be created by our various blocking rules and then differing our salting allocations accordingly?
> 
> So, if we had say:
> 
> ```
> blocking_rules_to_generate_predictions = [
> "l.col_a = r.col_a", 
> "l.col_b = r.col_b and l.col_c and r.col_c"
> ]
> ```
> 
> and we knew that `"l.col_a = r.col_a"` created 50,000 comparisons, whereas `"l.col_b = r.col_b and l.col_c and r.col_c"` only created 10,000, then we could distribute our salt partitions like so, to balance the load:
> 
> ```
> "l.col_a = r.col_a and __splink_salt = 1",
> ...
> "l.col_a = r.col_a and __splink_salt = 10",
> "l.col_b = r.col_b and l.col_c and r.col_c and __splink_salt = 1",
> "l.col_b = r.col_b and l.col_c and r.col_c and __splink_salt = 2"
> ```

For what it's worth, I have implemented exactly this (where N salts varies by rule) manually in my own work using large datasets (100s of millions of rows) and it works well.

**Comment by RobinL** [comment]
Closed by #278
